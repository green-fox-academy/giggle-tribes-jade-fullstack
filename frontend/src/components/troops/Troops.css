.component {
  grid-area: 3 / 1 / span 1 / span 2;
  display: grid;
  grid-template-columns: 140px 1fr;
  grid-template-rows: 140px 1fr;
  background-color: rgba(201, 204, 205, 0.8);
}
.component.image {
  margin: 20px;
  height: 100px;
  width: 100px;
  grid-area: 1 / 1 /1 /1;
  background-color: transparent;
}

.component.info {
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
  margin: 5px;
  grid-area: 2 / 1;
  background-color: transparent;
}

.component.info p {
  color: #676969;
}

.component.main {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  margin: 20px;
  grid-area: 1 / 2;
  background-color: transparent;
}

.component .button.image {
  width: 3em;
  height: 3em;
  background-color: transparent;
  box-shadow: none;
}
