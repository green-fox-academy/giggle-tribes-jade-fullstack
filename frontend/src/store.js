import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';

import { userReducer } from './reducers/userReducer';
import { kingdomReducer } from './reducers/kingdomReducer';
import { buildingsReducer } from './reducers/buildingsReducer';
import { resourcesReducer } from './reducers/resourcesReducer';
import { troopsReducer } from './reducers/troopsReducer';
import { tokenReducer } from './reducers/tokenReducer';
import { errorReducer } from './reducers/errorReducer';

function root(state = {}, action) {
  return {
    user: userReducer(state.user, action),
    kingdom: kingdomReducer(state.kingdom, action),
    buildings: buildingsReducer(state.buildings, action),
    resources: resourcesReducer(state.resources, action),
    troops: troopsReducer(state.troops, action),
    token: tokenReducer(state.token, action),
    error: errorReducer(state.error, action),
  };
}

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
const store = createStore(
  root,
  /* preloadedState, */ composeEnhancers(applyMiddleware(thunk))
);

export { store };
