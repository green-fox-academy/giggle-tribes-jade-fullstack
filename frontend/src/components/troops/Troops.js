import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import './Troops.css';
import { getTroopsAction } from '../../actions/TroopsActions';
import Button from './TroopButton';

import Troop from '../../assets/icons/troopbg.svg';

function Troops({ kingdom, troops, get }) {
  useEffect(() => {
    get(kingdom);
  }, [kingdom, get]);

  const analyseTroops = troops => {
    const result = {
      attack: 0,
      defence: 0,
      sustenance: 0,
      amountPerLevel: [],
    };
    troops.forEach(troop => {
      result.attack += troop.attack;
      result.defence += troop.defence;
      result.sustenance++;
      if (
        result.amountPerLevel.length === 0 ||
        !result.amountPerLevel.find(e => e.level === troop.level)
      ) {
        result.amountPerLevel.push({ level: troop.level, amount: 1 });
      } else {
        const index = result.amountPerLevel.findIndex(
          e => e.level === troop.level
        );
        result.amountPerLevel[index].amount++;
      }
    });
    return result;
  };

  const troopsSummary = analyseTroops(troops);

  return (
    <div className="troops component">
      <img className="troops component image" src={Troop} alt="troop icon" />
      <div className="troops component info">
        <p>{'Attack: ' + troopsSummary.attack}</p>
        <p>{'Defence: ' + troopsSummary.defence}</p>
        <p>{'Sustenance: ' + troopsSummary.sustenance}</p>
      </div>
      <div className="troops component main">
        {troopsSummary.amountPerLevel.map((troopSummary, i) => {
          return <Button key={i} troopSummary={troopSummary}></Button>;
        })}
      </div>
    </div>
  );
}

Troops.propTypes = {
  kingdom: PropTypes.number.isRequired,
  troops: PropTypes.array.isRequired,
};

const mapStateToProps = ({ kingdom, troops }) => {
  return { kingdom, troops };
};

const mapDispatchToProps = dispatch => {
  return {
    get: kingdomId => {
      dispatch(getTroopsAction(kingdomId));
    },
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Troops);
