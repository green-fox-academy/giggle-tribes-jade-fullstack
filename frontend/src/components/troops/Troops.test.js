import React from 'react';
import { Provider } from 'react-redux';
import { render, unmountComponentAtNode } from 'react-dom';
import { act, Simulate } from 'react-dom/test-utils';
import configureStore from 'redux-mock-store';

import Troops from './Troops';
import { store } from '../../store';

global.fetch = jest.fn(async () => {
  return await Promise.resolve({
    json: () =>
      Promise.resolve({
        troops: [
          {
            id: 1,
            level: 1,
            hp: 1,
            attack: 1,
            defence: 1,
            started_at: 12345789,
            finished_at: 12399999,
          },
          {
            id: 2,
            level: 1,
            hp: 1,
            attack: 1,
            defence: 1,
            started_at: 12345789,
            finished_at: 12399999,
          },
          {
            id: 3,
            level: 2,
            hp: 2,
            attack: 2,
            defence: 2,
            started_at: 12345789,
            finished_at: 12399999,
          },
        ],
      }),
  });
});

let container;
beforeEach(() => {
  fetch.mockClear();
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

it('should dispatch an action when rendered', async () => {
  const mockStore = configureStore([]);
  const mockedStore = mockStore({
    kingdom: 1,
    troops: [],
  });
  mockedStore.dispatch = jest.fn();
  await act(async () => {
    render(
      <Provider store={mockedStore}>
        <Troops />
      </Provider>,
      container
    );
  });
  expect(mockedStore.dispatch).toHaveBeenCalledTimes(1);
});

it('matches snapeshot', async () => {
  await act(async () => {
    render(
      <Provider store={store}>
        <Troops />
      </Provider>,
      container
    );
  });

  expect(container.innerHTML).toMatchSnapshot();
});
