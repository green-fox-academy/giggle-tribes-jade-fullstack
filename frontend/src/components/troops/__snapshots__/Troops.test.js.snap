// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapeshot 1`] = `"<div class=\\"troops component\\"><img class=\\"troops component image\\" src=\\"troopbg.svg\\" alt=\\"troop icon\\"><div class=\\"troops component info\\"><p>Attack: 4</p><p>Defence: 4</p><p>Sustenance: 3</p></div><div class=\\"troops component main\\"><div class=\\"troopsButton\\"><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-text troops button MuiButton-textSizeSmall MuiButton-sizeSmall\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"MuiButton-label\\"><img class=\\"troops button image\\" src=\\"troops.svg\\" alt=\\"troop icon\\"><p>2 Troop level 1</p></span><span class=\\"MuiTouchRipple-root\\"></span></button><div class=\\"costLabel\\"><p>Upgrade <br>10 <img class=\\"inlineIcon\\" src=\\"GoldIcon.svg\\" alt=\\"attack icon\\"> per troop</p></div><div></div></div><div class=\\"troopsButton\\"><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-text troops button MuiButton-textSizeSmall MuiButton-sizeSmall\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"MuiButton-label\\"><img class=\\"troops button image\\" src=\\"troops.svg\\" alt=\\"troop icon\\"><p>1 Troop level 2</p></span><span class=\\"MuiTouchRipple-root\\"></span></button><div class=\\"costLabel\\"><p>Upgrade <br>10 <img class=\\"inlineIcon\\" src=\\"GoldIcon.svg\\" alt=\\"attack icon\\"> per troop</p></div><div></div></div></div></div>"`;
